### ISSUE TITLE
go/types, types2: better error message for failing inference for arguments using type parameters

### ISSUE THREAD

**Issue by mpx** [golang/go#49800]
Using compiler from tip:
```
go version devel +45bae64015 Thu Nov 25 04:02:39 2021 +0000 linux/amd64
```

Calling a generic function from another using the same type parameters causes inference to fail. The compiler points to the function being called, not the location of the call:

```go
package main

import (
    "fmt"
    "sort"
)

func keys[V any, M ~map[string]V](m M) []string {
    out := make([]string, 0, len(m))
    for k := range m {
        out = append(out, k)
    }
    return out
}

func sortKeys[V any, M ~map[string]V](m M) []string {
    // Fails:
    ks := keys(m)
    // Works:
    //ks := keys[V, M](m)
    sort.Strings(ks)
    return ks
}

func main() {
    m := map[string]int{"foo": 1, "bar": 2}
    fmt.Println(keys(m))
}
```

Returns:

```
./keys.go:8:18: M does not match map[string]V
```

Ideally type checking should be able to infer the type parameters since they are identical, but I'm not sure if this is intended for Go 1.18.

Alternatively, there should to be a clearer error message pointing to the call location, not the callee.

I couldn't see a similar open bug, but this is complicated enough (for me) I might have missed it.

Cc @griesemer @findleyr

**Comment by hanchaoqun** [comment]
Duplicate with https://github.com/golang/go/issues/49575, and the CL https://go-review.googlesource.com/c/go/+/362776

**Comment by mpx** [comment]
Thanks, yes, appears to be a duplicate of #49575, but changing the inference rules is pushed to Go 1.19.

I think the error message should be fixed in Go 1.18 - it needs to point to the call site and should have a clearer description.

**Comment by hanchaoqun** [comment]
> Thanks, yes, appears to be a duplicate of #49575, but changing the inference rules is pushed to Go 1.19.
> 
> I think the error message should be fixed in Go 1.18 - it needs to point to the call site and should have a clearer description.

@mpx  Maybe this CL https://golang.org/cl/362634 can fix the error report on the wrong line.
